### DISCUSSION TITLE
When to use "deep copy"? (Doc string is over my head)

### DISCUSSION THREAD

**Discussion by michaelaye** [pydata/xarray#4962] [Q&A] [answered]
[Dask delayed best practices](https://docs.dask.org/en/latest/delayed-best-practices.html) advises to not mutate functional inputs.
As I'm pushing small xarray.DataArrays around that I change in the computation, I now understand I should copy them.
But I cannot find any advise on if that should be a deep copy or not? 
The docstring for the deep copy option is frankly over my head. I'm just a poor scientist that for a good reason changed from C++ to Python all those years ago, I don't want to manage memory, I want to look at my data. :)  ( I figured in the discussion part of xarray I could add such a comment without sounding obnoxious, I wouldn't do so in an issue.
Any advise on this appreciated?

**Comment by max-sixty** [comment, marked as answer]
Others can give more precise definitions but my mental model:
- `.deepcopy` copies everything that the object references. It shares nothing with the existing object.
- `.copy` only copies the container. Anything that the existing object referenced will still be referenced by the new object.

One way to see this is to try `copy` and `deepcopy`-ing a dataset and see how these two respective operations on the copy are affected:
- Assign an `attrs`
- Mutate the underlying array; i.e. `ds["lat"][0] = 1`

**Reply by michaelaye** [reply]
Thanks for the reply!

I could not find any change propagation to any direction, should I

Here's what I did:

```python
pix = get_example_pix()
```
![image](https://user-images.githubusercontent.com/69774/109531512-37bfa180-7a75-11eb-8003-044c7ed00dff.png)
```python
pix.attrs['test'] = "Test"
shallow = pix.copy()
deep = pix.copy(deep=True)
```
I then went ahead and:
* changed lat and changed attrs["test"] to "Test2" in one object and checked the other 2 for changes
* Did this with fresh copies each time, 3 times, by either changing something in `pix`, `shallow`, or `deep`, but haven't seen any change propagation.

Am I doing this right? For a deeper understanding it might help what precisely is `the container` you mentioned above? Which parts of an xarray.Dataset are mere "references" ?

**Reply by keewis** [reply]
I think the only difference between a deep and a shallow copy is that the data is copied, too:

<details><summary>difference between shallow and deep copy</summary>

```python
In [4]: import xarray as xr
   ...: 
   ...: ds = xr.Dataset({"a": ("x", ["a", "b", "c"])}, attrs={"attr": "value"})
   ...: shallow_copy = ds.copy(deep=False)
   ...: deep_copy = ds.copy(deep=True)
   ...: 
   ...: print("original dataset:", ds, sep="\n")
   ...: 
   ...: ds.attrs["attr"] = "other_value"
   ...: ds.a[0] = "z"
   ...: 
   ...: print("modified dataset:", ds, sep="\n")
   ...: print("shallow copy:", shallow_copy, sep="\n")
   ...: print("deep copy:", deep_copy, sep="\n")
original dataset:
<xarray.Dataset>
Dimensions:  (x: 3)
Dimensions without coordinates: x
Data variables:
    a        (x) <U1 'a' 'b' 'c'
Attributes:
    attr:     value
modified dataset:
<xarray.Dataset>
Dimensions:  (x: 3)
Dimensions without coordinates: x
Data variables:
    a        (x) <U1 'z' 'b' 'c'
Attributes:
    attr:     other_value
shallow copy:
<xarray.Dataset>
Dimensions:  (x: 3)
Dimensions without coordinates: x
Data variables:
    a        (x) <U1 'z' 'b' 'c'
Attributes:
    attr:     value
deep copy:
<xarray.Dataset>
Dimensions:  (x: 3)
Dimensions without coordinates: x
Data variables:
    a        (x) <U1 'a' 'b' 'c'
Attributes:
    attr:     value
```

</details>

so the data is referenced in shallow copies while the `attrs` and other metadata are copied

**Reply by max-sixty** [reply]
Thanks @keewis , my comment re "assign an `attr`" was incorrect, appreciate you correcting me!

As you say — the important difference is whether the underlying data is copied. "Mutate the underlying array; i.e. `ds["lat"][0] = 1`" will show that.

**Reply by keewis** [reply]
actually, you were right about `attrs` also being affected by this:

<details><summary><tt>attrs</tt> copy</summary>

```python
In [3]: import xarray as xr
   ...: 
   ...: ds = xr.Dataset({"a": ("x", ["a", "b", "c"])}, attrs={"attr": ["a", "b"]})
   ...: shallow_copy = ds.copy(deep=False)
   ...: deep_copy = ds.copy(deep=True)
   ...: 
   ...: print("original dataset:", ds, sep="\n")
   ...: 
   ...: ds.attrs["attr"].append("c")
   ...: ds.a[0] = "z"
   ...: 
   ...: print("modified dataset:", ds, sep="\n")
   ...: print("shallow copy:", shallow_copy, sep="\n")
   ...: print("deep copy:", deep_copy, sep="\n")
original dataset:
<xarray.Dataset>
Dimensions:  (x: 3)
Dimensions without coordinates: x
Data variables:
    a        (x) <U1 'a' 'b' 'c'
Attributes:
    attr:     ['a', 'b']
modified dataset:
<xarray.Dataset>
Dimensions:  (x: 3)
Dimensions without coordinates: x
Data variables:
    a        (x) <U1 'z' 'b' 'c'
Attributes:
    attr:     ['a', 'b', 'c']
shallow copy:
<xarray.Dataset>
Dimensions:  (x: 3)
Dimensions without coordinates: x
Data variables:
    a        (x) <U1 'z' 'b' 'c'
Attributes:
    attr:     ['a', 'b', 'c']
deep copy:
<xarray.Dataset>
Dimensions:  (x: 3)
Dimensions without coordinates: x
Data variables:
    a        (x) <U1 'a' 'b' 'c'
Attributes:
    attr:     ['a', 'b']
```

</details>

**Reply by max-sixty** [reply]
You're insufficiently critical of me @keewis ! 😄 

In order to see the difference between copy & deepcopy in `attrs`, we have to mutate the value of an attr. Adding an attr doesn't show any difference. 

In that way, a `Dataset` is not the same as a python object like a list or dict. That's the element I think you corrected me on.

Do you agree?

**Reply by keewis** [reply]
my understanding of `.copy` is that it applies `copy.copy` / `copy.deepcopy` to whatever it is wrapping, so what exactly happens is determined by their implementation of `__copy__` / `__deepcopy__`. `dict` and `ndarray` happen to reference the data if they are shallow-copied, but other types (i.e. duckarray objects) might do something different.

**Reply by dcherian** [reply]
(we should add this to the docstring)
